src/main/java/org/breedinginsight/daos/ProgramLocationDAO.java

```diff
@@ -132,17 +139,44 @@ private List<ProgramLocation> parseRecords(List<Record> records) {
             location.setTopography(Topography.parseSQLRecord(record));
             location.setCreatedByUser(org.breedinginsight.model.User.parseSQLRecord(record, createdByUser));
             location.setUpdatedByUser(User.parseSQLRecord(record, updatedByUser));
-            resultLocations.add(location);
+            resultLocations.put(location.getId(), location);
+            locationIds.add(location.getId());
         }
 
-        return resultLocations;
+        if(fetchBrAPIObject && !resultLocations.isEmpty()) {
+            List<BrAPILocation> brAPILocations = getBrapiLocations(locationIds, resultLocations.values().stream().findFirst().get().getProgramId());
+
+            if (brAPILocations.size() != resultLocations.size()) {
+                throw new IllegalStateException("Did not find BrAPI Location objects for each location");
+            } else {
+                brAPILocations.forEach(brapiLocation -> {
+                    Optional<BrAPIExternalReference> externalReference = Utilities.getExternalReference(brapiLocation.getExternalReferences(), referenceSource);
+                    if (externalReference.isPresent()) {
```

Comment:
This if-else can be eliminated with `BrAPIExternalReference externalReference = Utilities.getExternalReference(...).orElseThrow(() -> new IllegalStateException(...))`